src/core/qgstracer.cpp

```diff
@@ -0,0 +1,676 @@
+  g.inactiveEdges << eIdx;
+
+  // add new vertex and edges to the graph
+  g.v.append( v );
+  g.e.append( e1 );
+  g.e.append( e2 );
+  g.joinedVertices++;
+
+  return vIdx;
+}
+
+
+int point_in_graph( QgsTracerGraph& g, const QgsPoint& pt )
+{
+  // try to use existing vertex in the graph
+  int v = point2vertex( g, pt );
+  if ( v != -1 )
+    return v;
+
+  // try to add the vertex to an edge (may fail if point is not on edge)
+  return join_vertex_to_graph( g, pt );
+}
+
+
+void reset_graph( QgsTracerGraph& g )
+{
+  // remove extra vertices and edges
+  g.v.resize( g.v.count() - g.joinedVertices );
+  g.e.resize( g.e.count() - g.joinedVertices * 2 );
+  g.joinedVertices = 0;
+
+  // fix vertices of deactivated edges
+  foreach ( int eIdx, g.inactiveEdges )
+  {
+    if ( eIdx >= g.e.count() )
+      continue;
+    const QgsTracerGraph::E& e = g.e[eIdx];
+    QgsTracerGraph::V& v1 = g.v[e.v1];
+    for ( int i = 0; i < v1.edges.count(); ++i )
+    {
+      if ( v1.edges[i] >= g.e.count() )
+        v1.edges.remove( i-- );
+    }
+    v1.edges << eIdx;
+
+    QgsTracerGraph::V& v2 = g.v[e.v2];
+    for ( int i = 0; i < v2.edges.count(); ++i )
+    {
+      if ( v2.edges[i] >= g.e.count() )
+        v2.edges.remove( i-- );
+    }
+    v2.edges << eIdx;
+  }
+
+  g.inactiveEdges.clear();
+}
+
+
+void extract_linework( QgsGeometry* g, QgsMultiPolyline& mpl )
+{
+  switch ( QgsWKBTypes::flatType( g->geometry()->wkbType() ) )
+  {
+    case QgsWKBTypes::LineString:
+      mpl << g->asPolyline();
+      break;
+
+    case QgsWKBTypes::Polygon:
+      foreach ( const QgsPolyline& ring, g->asPolygon() )
+        mpl << ring;
+      break;
+
+    case QgsWKBTypes::MultiLineString:
+      foreach ( const QgsPolyline& linestring, g->asMultiPolyline() )
+        mpl << linestring;
+      break;
+
+    case QgsWKBTypes::MultiPolygon:
+      foreach ( const QgsPolygon& polygon, g->asMultiPolygon() )
+        foreach ( const QgsPolyline& ring, polygon )
+          mpl << ring;
+      break;
+
+    default:
+      break;  // unkown type - do nothing
+  }
+}
+
+// -------------
+
+
+QgsTracer::QgsTracer()
+    : mGraph( 0 )
+    , mMaxFeatureCount( 0 )
+{
+}
+
+
+bool QgsTracer::initGraph()
+{
+  if ( mGraph )
+    return true; // already initialized
+
+  QgsFeature f;
+  QgsMultiPolyline mpl;
+
+  // extract linestrings
+
+  // TODO: use QgsPointLocator as a source for the linework
+
+  QTime t1, t2, t2a, t3;
+
+  t1.start();
+  int featuresCounted = 0;
+  foreach ( QgsVectorLayer* vl, mLayers )
+  {
+    QgsCoordinateTransform ct( vl->crs(), mCRS );
+
+    QgsFeatureRequest request;
+    request.setSubsetOfAttributes( QgsAttributeList() );
+    if ( !mExtent.isEmpty() )
+      request.setFilterRect( ct.transformBoundingBox( mExtent, QgsCoordinateTransform::ReverseTransform ) );
+
+    QgsFeatureIterator fi = vl->getFeatures( request );
+    while ( fi.nextFeature( f ) )
+    {
+      if ( !f.geometry() )
```

Comment:
Well, if the transform isn't required then you'll only need to use constGeometry calls in lines 473 and 488. It could potentially save QgsFeature detachments when the transform isn't needed.